### PR TITLE
Make typing strict

### PR DESCRIPTION
**Pull request by dstansby**
This enables strict mypy typing, fixing (almost) https://github.com/zarr-developers/zarr-python/issues/1783 🎉 

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by jhamman** [comment]
@dstansby - let us know when this is ready for a review.

**Comment by dstansby** [comment]
👍 should be good to review now

File: src/zarr/config.py [hunk 1]
```diff
@@ -18,6 +18,6 @@
 
 def parse_indexing_order(data: Any) -> Literal["C", "F"]:
```
Comment (jhamman) [inline src/zarr/config.py:19]: noting that this function also exists in `zarr.common`... No need to fix here but we don't need both.

**Review by jhamman — APPROVED** [review]
Thanks @dstansby!

**Pull request merged**